### PR TITLE
Consolidate code for equals/hashCode testing in central utility class

### PR DESCRIPTION
**Pull request by cbuescher**
Currently test that check that equals() and hashCode() are working as expected for classes implementing them are quiet similar. This change moves common assertions in this method to a common utility class. In addition, another common utility function in most of these test classes that creates copies of input object by running them through a StreamOutput and reading them back in, is moved to ESTestCase so it can be shared across all these classes.

Relates to #20629

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java [hunk 1]
```diff
@@ -41,49 +47,54 @@ public static FuzzyOptions randomFuzzyOptions() {
...
-        builder.setFuzziness(original.getEditDistance())
-            .setFuzzyPrefixLength(original.getFuzzyPrefixLength())
-            .setFuzzyMinLength(original.getFuzzyMinLength())
-            .setMaxDeterminizedStates(original.getMaxDeterminizedStates())
-            .setTranspositions(original.isTranspositions())
-            .setUnicodeAware(original.isUnicodeAware());
+        builder.setFuzziness(original.getEditDistance()).setFuzzyPrefixLength(original.getFuzzyPrefixLength())
+                .setFuzzyMinLength(original.getFuzzyMinLength()).setMaxDeterminizedStates(original.getMaxDeterminizedStates())
+                .setTranspositions(original.isTranspositions()).setUnicodeAware(original.isUnicodeAware());
         switch (randomIntBetween(0, 5)) {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java]: I've started to appreciate
```
Runnable r = randomFrom(
 () -> builder.setFuzziness....,
 () -> builder.asdfadfasdf
);
r.run();
```

That way you can't forget to change the upper bound when you add a new random option.

File: core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java [hunk 2]
```diff
@@ -41,49 +47,54 @@ public static FuzzyOptions randomFuzzyOptions() {
...
     }
 
-    @Override
-    protected FuzzyOptions readFrom(StreamInput in) throws IOException {
-        return new FuzzyOptions(in);
+    /**
+     * Test serialization and deserialization
+     */
+    public void testSerialization() throws IOException {
+        for (int i = 0; i < NUMBER_OF_RUNS; i++) {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java]: I think we've been moving away from this sort of "poor man's repeat" and just doing the run once and hoping for CI to catch enough stuff. If we really aren't sure that is enough maybe we should add a couple of specific mutations to test every time?

**Review by nik9000 — APPROVED** [review]
I think this is a good cleanup in general. I left a couple of specific

File: core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java (hunk 1 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java]: Thats a great suggestion, will change this here and continue to use this in similar cases when I see it.

File: core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java (hunk 2 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/completion/FuzzyOptionsTests.java]: I understand the concern, but CI runs are so infrequent now really and also if I run those tests from my IDE I usually really want to test more than one random permutation of the objects under test. This has caught so many things for me already in the past that I think its really worth it. Plus, this doesn't really add a lot of overhead, last time I looked at it these additional runs added almost nothing to the runtime of the test compared to initiallization. So to summarize: I think it buys a lot more to have more than one permutation per test than it costs.

**Comment by nik9000** [comment]
I understand the multiple tests per iteration. Let's get it in now and open
an issue to talk about it. The way it works now is less good than `@Repeat`
because you can't select the failing case and rerun in. I know we have good
reasons not to commit repeat, but it seems lame to reimplement it to work
around forbidden APIs.

**Comment by cbuescher** [comment]
@nik9000 thanks. Re: your comment about reproducability: I never had any problem with that. With the right seed, test fail predictably for the nth run. If you want to find out what went wrong with that one and you need debugging you can break on condition or sth. else.

**Comment by cbuescher** [comment]
Merged to 5.x with 5275860cbc38254d8dc8d18e74e23e86e913bc11

**Pull request merged**